### DISCUSSION TITLE
pcp_combine with -add

### DISCUSSION THREAD

**Discussion by Jeff-Duda** [dtcenter/METplus#2712] [Configuration] [answered]
What am I doing wrong?

When entered in a bash script, the output from pcp_combine is this:

```
DEBUG 1: Start pcp_combine by role.amb-verif(5091) at 2024-10-02 05:44:24Z  cmd: pcp_combine -add /mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427600000003 -field name="APCP"; level="A03"; /mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427600000006 -field name="APCP"; level="A03"; /lfs5/BMC/amb-verif/RT_ensemble_products/temp_products/NAMnest/2024100200/NAMnest_precip_06h_f06.nc
DEBUG 2: Since the "-field" command line option was used, parsing the command line arguments as a list of files.
DEBUG 2: Performing derivation command (sum) for 4 files.
DEBUG 1: Reading data (name="APCP";) from input file: /mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427600000003
ERROR  :
ERROR  : VarInfo::set_level_info_grib() - either level or GRIB_lvl_typ, GRIB_lvl_val1 and GRIB_lvl_val2 (if necessary) must be specified in field information
ERROR  :
```

Whereas if I enter essentially the same command right on the command line, it seems to work:
```
pcp_combine -add /mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427600000003 -field 'name="APCP"; level="A03";' /mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427600000006 -field 'name="APCP"; level="A03";' totallynewfile.nc
DEBUG 1: Start pcp_combine by role.amb-verif(5091) at 2024-10-02 05:52:52Z  cmd: pcp_combine -add /mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427600000003 -field name="APCP"; level="A03"; /mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427600000006 -field name="APCP"; level="A03"; totallynewfile.nc
DEBUG 2: Since the "-field" command line option was used, parsing the command line arguments as a list of files.
DEBUG 2: Performing derivation command (sum) for 2 files.
DEBUG 1: Reading data (name="APCP"; level="A03";) from input file: /mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427600000003
DEBUG 1: Reading data (name="APCP"; level="A03";) from input file: /mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427600000006
DEBUG 2: Skipping 91 of 1905141 grid points which do not meet the valid data threshold (1).
DEBUG 1: Creating output file: totallynewfile.nc
DEBUG 2: Writing output variable "APCP_06" for the "sum" of "APCP/A03".
DEBUG 2: Performing derivation command (sum) for 2 files.
DEBUG 1: Reading data (name="APCP"; level="A03";) from input file: /mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427600000003
DEBUG 1: Reading data (name="APCP"; level="A03";) from input file: /mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427600000006
DEBUG 2: Skipping 91 of 1905141 grid points which do not meet the valid data threshold (1).
DEBUG 2: Writing output variable "APCP_06" for the "sum" of "APCP/A03".
terminate called after throwing an instance of 'netCDF::exceptions::NcNameInUse'
  what():  NetCDF: String match to name in use
file: ncGroup.cpp  line:496
FATAL: Received Signal Abort. Exiting 6
```

Note that despite the "FATAL" wording at the bottom, the expected output file is present and has a full ncdump:
```
(avid_verify) 532 role.amb-verif@fe7:/lfs5/BMC/amb-verif/RT_ensemble_products/log$ ncdump -h totallynewfile.nc
netcdf totallynewfile {
dimensions:
        lat = 1059 ;
        lon = 1799 ;
variables:
        float lat(lat, lon) ;
                lat:long_name = "latitude" ;
                lat:units = "degrees_north" ;
                lat:standard_name = "latitude" ;
        float lon(lat, lon) ;
                lon:long_name = "longitude" ;
                lon:units = "degrees_east" ;
                lon:standard_name = "longitude" ;
        float APCP_06(lat, lon) ;
                APCP_06:coordinates = "lat lon" ;
                APCP_06:name = "APCP_06" ;
                APCP_06:long_name = "Total Precipitation" ;
                APCP_06:level = "A6" ;
                APCP_06:units = "kg/m^2" ;
                APCP_06:_FillValue = -9999.f ;
                APCP_06:init_time = "20241002_000000" ;
                APCP_06:init_time_ut = "1727827200" ;
                APCP_06:valid_time = "20241002_060000" ;
                APCP_06:valid_time_ut = "1727848800" ;
                APCP_06:accum_time = "060000" ;
                APCP_06:accum_time_sec = 21600 ;

// global attributes:
                :FileOrigins = "File totallynewfile.nc generated 20241002_055255 UTC on host fe7 by the MET pcp_combine tool" ;
                :MET_version = "V11.1.0" ;
                :MET_tool = "pcp_combine" ;
                :RunCommand = "Addition: 2 files." ;
                :Projection = "Lambert Conformal" ;
                :hemisphere = "N" ;
                :scale_lat_1 = "38.500000" ;
                :scale_lat_2 = "38.500000" ;
                :lat_pin = "21.138000" ;
                :lon_pin = "-122.720000" ;
                :x_pin = "0.000000" ;
                :y_pin = "0.000000" ;
                :lon_orient = "-97.500000" ;
                :d_km = "3.000000" ;
                :r_km = "6371.200000" ;
                :nx = "1799" ;
                :ny = "1059 grid_points" ;
```

**Comment by JohnHalleyGotway** [comment]
Hi @Jeff-Duda, I see you're having difficulty with the `-add` option of the `pcp_combine` tool. The issue is just in how your command line is structured and is easily fixed. But I can see what led to the confusion.

Here's a selection taken from the [usage statement](https://metplus.readthedocs.io/projects/met/en/latest/Users_Guide/reformat_grid.html#pcp-combine-usage) for pcp_combine:
```
Usage: pcp_combine
	[-sum] sum_args |
	-add input_files |
	-subtract input_files |
	-derive stat_list input_files
	out_file
...
where
...
	INPUT_FILES:
		file_1 config_str_1 ... file_n config_str_n | 
		file_1 ... file_n |
		input_file_list
```
Note that there are 3 supported ways of defining the `INPUT_FILES`:
1. `file_1 config_str_1 ... file_n config_str_n` -> For each input file listed on the command line, follow it with a configuration string to define the data to be read from that file. Note that these `config_str_n` strings DO NOT include the `-config` option... which is likely the source of the confusion.
2. `file_1 ... file_n` -> For each input file listed on the command line, read the SAME field(s) of data from all of them and specify the field(s) using the `-field` option one or more times.
3. `input_file_list` -> Same as (2), but the input files are read from an ASCII file list rather than the command line.

Since you're reading the same field (`'name="APCP"; level="A03";'`) from both input files, you could technically format it any of these 3 ways, but here's how I'd probably do it...

Option (2) - to read the same field from both input files:
```
pcp_combine -add \
/mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427600000003 \
/mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427600000006 \
-field 'name="APCP"; level="A03";' \
totallynewfile.nc
```
Note that you got that NetCDF error message (`netCDF::exceptions::NcNameInUse`) because you'd used the `-field` option twice, both times requesting the same input data. So `pcp_combine` performed the same operation twice and attempted to write the same output variable name twice... which led to that error about output variable name already existing.

FYI, here's what the corresponding formatting would look like for options (1) and (3) with your data:

Option (1) - useful when the input field names differ:
```
pcp_combine -add \
/mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427600000003 'name="APCP"; level="A03";' \
/mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427600000006 'name="APCP"; level="A03";' \
totallynewfile.nc
```
Option (3) - useful when you have a long list of input files:
```
echo "/mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427600000003 /mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427600000006" > my_list_of_files
pcp_combine -add my_list_of_files -field 'name="APCP"; level="A03";' totallynewfile.nc
```
All 3 of these options will produce the same output.

Hope that helps!

**Reply by Jeff-Duda** [reply]
Thank you for the prompt response, John. I will add that I am using METv11.1.0 for this.

Here is the problem when I use your suggestion for option (1):
`pcp_combine -add  /mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427612000003 'name="APCP"; level="A03";' /mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427612000006 'name="APCP"; level="A03";' /lfs5/BMC/amb-verif/RT_ensemble_products/temp_products/NAMnest/2024100212/NAMnest_precip_06h_f06.nc`

Which is equivalent to
`pcp_combine -add \`
`/mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427612000003 'name="APCP"; level="A03";' \`
`/mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427612000006 'name="APCP"; level="A03";' \`
`/lfs5/BMC/amb-verif/RT_ensemble_products/temp_products/NAMnest/2024100212/NAMnest_precip_06h_f06.nc`

I get this error statement from pcp_combine:

> DEBUG 1: Start pcp_combine by role.amb-verif(5091) at 2024-10-02 16:39:55Z  cmd: .....(yadayadayada)
> DEBUG 2: Since the "-field" command line option was not used, parsing the command line arguments a list of files, each followed by a configuration string.
> ERROR  :
> ERROR  : process_add_sub_derive_args() -> file name used when config string expected (/mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427612000006). Did you forget the "-field" option?
> ERROR  :

If I try your suggestion for option (2) I get a different error:
`pcp_combine -add  /mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427612000003 /mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427612000006 -field 'name="APCP"; level="A03";' /lfs5/BMC/amb-verif/RT_ensemble_products/temp_products/NAMnest/2024100212/NAMnest_precip_06h_f06.nc`

which is equivalent to
`pcp_combine -add  \`
`/mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427612000003 \`
`/mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427612000006 \`
`-field 'name="APCP"; level="A03";' \`
`/lfs5/BMC/amb-verif/RT_ensemble_products/temp_products/NAMnest/2024100212/NAMnest_precip_06h_f06.nc`

The output is:

> DEBUG 1: Start pcp_combine by role.amb-verif(5091) at 2024-10-02 16:46:23Z  cmd: ......yadayadayada
> DEBUG 2: Since the "-field" command line option was used, parsing the command line arguments as a list of files.
> DEBUG 2: Performing derivation command (sum) for 2 files.
> DEBUG 1: Reading data (name=\"APCP\"; level=\"A03\";) from input file: /mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427612000003
> ERROR  :
> ERROR  : yyerror() -> syntax error in file "/tmp/met_config_3502907_1"
> ERROR  :
> ERROR  :    line   = 1
> ERROR  :
> ERROR  :    column = -15
> ERROR  :
> ERROR  :    text   = "APCP"; level="A03";
> ERROR  : "
> ERROR  :
> ERROR  :
> ERROR  : name=\"APCP\"; level=\"A03\";
> ERROR  : ^^^^_________________________
> ERROR  :
> 

So I need further help on this. Thanks.

**Reply by JohnHalleyGotway** [reply, marked as answer]
@Jeff-Duda, hmm, I'm not positive on this, but my *guess* is that perhaps you're copying the command directly from your mail browser and some hidden characters are unwittingly being included which cause parsing problems? I've had something similar happen to me in the past. When I view this discussion on GitHub, I am able to copy/paste without any hidden characters wreaking havoc.

I did log on to jet and ran the commands listed below, and they all run without error. Note that I wrote to my current directory instead of your project directory.
```
module load intel/2022.1.2
module load contrib
module use /contrib/met/modulefiles
module load met/11.1.0
which pcp_combine
pcp_combine -add  /mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427612000003 /mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427612000006 -field 'name="APCP"; level="A03";' NAMnest_precip_06h_f06.nc
pcp_combine -add /mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427612000003 /mnt/lfs5/BMC/public/data/grib/ftp/7/0/84/0_1905141_30/2427612000006 -field 'name="APCP"; level="A03";' NAMnest_precip_06h_f06.nc
```
You can also see these in `/home/John.H.Gotway/metplus_discussion_2712_commands.sh`.

**Reply by Jeff-Duda** [reply]
@JohnHalleyGotway it was probably the case that my script code is reading single and double quotes differently than I was expecting for typical bash behavior. I finally got it to work from my script by changing the quoting. pcp_combine overall seems extremely sensitive to quoting.

**Reply by JohnHalleyGotway** [reply]
Jeff, yes, that is unfortunately true. Each configuration string should be fully enclosed in single quotes and then strings inside should be done with double quotes. And yes, getting the syntax exactly right in shell scripts can be maddening.

I'll note that this is one advantage of running MET via the METplus Python wrappers. They handle the syntax details.
